### PR TITLE
MSC1954: Proposal to remove prev_content from the essential keys list

### PR DESCRIPTION
**Pull request by neilisfragile**
[Rendered](https://github.com/matrix-org/matrix-doc/blob/neilj/Remove-prev_content-from-the-essential-keys-list2/proposals/1954-remove-prev_event-from-essential-keys-list.md)

### CODE REVIEW

File: proposals/1954-remove-prev_event-from-essential-keys-list.md [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+retrospectively to describe the de-facto redaction algorithm.
+
+## Potential issues
+
+It is theoretically possible that a closed federation could exist whose servers
+do follow the spec as is. This MSC would render those servers uncompliant with
+the spec. On balance this seems unlikely and in the worst case those
+implementors could add the change to a subsequent room version, eventually
+reaching spec consistency as older room versions are deprecated.
+
```
Comment (KitsuneRal) [inline proposals/1954-remove-prev_event-from-essential-keys-list.md]: Probably worth adding:
Clients that redact events according to the spec as is keep `prev_content` and by that _already_ have their representation of redacted events diverting from that on the server(s). Client authors will have to update their code to drop `prev_content` - however, given that `prev_content` should not be used in important calculation/visualisation anyway, this looks to be relatively small impact compared to introducing the new room version as described above.

File: proposals/1954-remove-prev_event-from-essential-keys-list.md [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+## Security considerations
+
+I am unaware of any security issues related to this proposal, but can certainly
+see issues with a precedent that the federation deviates from the spec.
```
Comment (KitsuneRal) [inline proposals/1954-remove-prev_event-from-essential-keys-list.md]: Actually, keeping `prev_content` as the current text of the spec says carries a security implication: if somebody stuffed, e.g., illegal or abusive data into a state event and then overwritten the same state event once, those data would be carefully preserved and disseminated via `prev_content` even after redaction (if implemented as-per-spec). Therefore `prev_content` is not just inessential, it's harmful if kept by redaction. Not sure if we should add it specifically here but it's probably worth stating at least.

**Comment by turt2live** [comment]
in an attempt to help @neilisfragile get this through:

@mscbot fcp merge

**Comment by dbkr** [comment]
Seems entirely reasonable

File: proposals/1954-remove-prev_event-from-essential-keys-list.md [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+the spec. On balance this seems unlikely and in the worst case those
+implementors could add the change to a subsequent room version, eventually
+reaching spec consistency as older room versions are deprecated.
+
+Another scenario is that a client may redact events according to the spec as is
+and persist prev_content through the redaction, thereby diverting from that on
+the server(s). Client authors will have to update their code to drop
+```prev_content``` - however, given that prev_content should not be used in
+important calculations and/or visualisations, this ought to be relatively
+uninvaisive change.
```
Comment (KitsuneRal) [inline proposals/1954-remove-prev_event-from-essential-keys-list.md]:
```suggestion
uninvasive change.
```

Comment (anoadragon453) [reply proposals/1954-remove-prev_event-from-essential-keys-list.md]: noninvasive

File: proposals/1954-remove-prev_event-from-essential-keys-list.md [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+be spec compliant.
+
+As a result it seems pragmatic to have the spec reflect reality, acknowledge
+that the spec and federation have deviated and instead update the spec
+retrospectively to describe the de-facto redaction algorithm.
+
+## Potential issues
+
+It is theoretically possible that a closed federation could exist whose servers
+do follow the spec as is. This MSC would render those servers uncompliant with
```
Comment (anoadragon453) [inline proposals/1954-remove-prev_event-from-essential-keys-list.md]:
```suggestion
do follow the spec as is. This MSC would render those servers incompliant with
```

**Comment by anoadragon453** [comment]
Makes sense to me.

**Comment by turt2live** [comment]
This doesn't need an implementation because the implementation is already correct. The spec is wrong in this scenario. Flagging it as spec-pr-missing and spec-bug due to the nature of the problem.

**Comment by turt2live** [comment]
Spec PR: https://github.com/matrix-org/matrix-doc/pull/2016

**Comment by turt2live** [comment]
merged 🎉

**Pull request merged**